### PR TITLE
Document how to write private code and util/utils.py

### PR DESCRIPTION
**Pull request by dstansby**
I think this covers what we agreed on the call today? Opinions/suggestions for changes and additions still very welcome.

### CODE REVIEW

File: docs/dev_guide/contents/code_standards.rst [hunk 1]
```diff
@@ -49,6 +49,13 @@ Coding Style/Conventions
 
 * General utilities necessary for but not specific to the package should be placed in the ``sunpy.utils`` module.
```
Comment (ayshih) [inline docs/dev_guide/contents/code_standards.rst:50]:
```suggestion
* General utilities necessary for but not specific to the package should be placed in the ``sunpy.util`` module.
```

File: docs/dev_guide/contents/code_standards.rst [hunk 2]
```diff
@@ -49,6 +49,13 @@ Coding Style/Conventions
...
 
+Private code
+============
+Any classes, functions, or variables that are private (ie. not part of the user facing API) should always be pre-pended with an underscore and should live in:
+
+- a sub-package, if it is only used within that sub-package.
+- ``sunpy.utils`` if it is used across multiple sub-packages.
```
Comment (ayshih) [inline docs/dev_guide/contents/code_standards.rst:57]:
```suggestion
- ``sunpy.util`` if it is used across multiple sub-packages.
```

**Comment by ayshih** [comment]
Hmm.  Would such guidance mean that the transformation functions in `sunpy.coordinates.transformations`, none of which are directly called by users, should be prepended with an underscore?  Would such guidance mean that we need to start prepending underscores to everything in `sunpy.util`?

**Comment by dstansby** [comment]
Yes, it would. My own thinking behind this is:
- Underscored stuff doesn't show up in tab auto-completion (e.g. in a jupyter notebook)
- If you're using `dir()` to see what's available in a module, an underscore makes it clear that certain functions/classes/methods are not intended for user use

**Comment by Cadair** [comment]
I am not sure I am in favour of a policy which is this blunt. I see the appeal, but at the same time it feels redundant to not have a way to signal "this whole file isn't public API".

One potential solution could be to rename `sunpy/coordinates/transformations.py` to `sunpy/coordinates/_transformations.py` but that too feels weird.

**Comment by Cadair** [comment]
Yeah I was mostly talking about the underscore thing. I think it's probably worth sorting out some way of indicating that certain sub-subpackages are not subject to stability promises though.

**Comment by dstansby** [comment]
How about now? I've removed underscoring. Because this means it's unclear what denotes private/public API in sub-packages, I've suggested normalising this to living in `sunpy.{subpackage}.utils.py`.

**Comment by nabobalis** [comment]
Do we want to talk about the future of utils and if we want to actually make that develop facing and therefore we will maintain its API like we do for user code?

**Comment by dstansby** [comment]
I've lost track of what we decided (if anything). If someone else wants to take my branch and run feel free.

**Comment by nabobalis** [comment]
> I've lost track of what we decided (if anything). If someone else wants to take my branch and run feel free.

I've pushed some changes. Lets see how it goes.

File: docs/dev_guide/contents/code_standards.rst [hunk 3]
```diff
@@ -49,6 +49,29 @@ Coding Style/Conventions
 
 * General utilities necessary for but not specific to the package should be placed in the ``sunpy.utils`` module.
```
Comment (nabobalis) [inline docs/dev_guide/contents/code_standards.rst]:
```suggestion
```

File: docs/dev_guide/contents/code_standards.rst [hunk 4]
```diff
@@ -47,7 +47,30 @@ Coding Style/Conventions
 
 * ``__init__.py`` files for modules should not contain any significant implementation code. ``__init__.py`` can contain docstrings and code for organizing the module layout.
 
-* General utilities necessary for but not specific to the package should be placed in the ``sunpy.utils`` module.
+* General utilities necessary for but not specific to the package should be placed in the ``sunpy.util`` module.
```
Comment (nabobalis) [inline docs/dev_guide/contents/code_standards.rst:50]:
```suggestion
```

**Review by Cadair — APPROVED** [review]

**Pull request merged**